File: internal/bot/command/add_apple_app.go
package command

import (
	"fmt"
	"strconv"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"github.com/miti99/store-scraper-bot-go/internal/api/apple"
	"github.com/miti99/store-scraper-bot-go/internal/api/apple/request"
	"github.com/miti99/store-scraper-bot-go/internal/config"
	"github.com/miti99/store-scraper-bot-go/internal/repository"
)

// /addapple <id|appId> [country=vn] — Java AddAppleAppCommand.
type AddAppleAppCommand struct {
	cfg       *config.Config
	adminRepo *repository.AdminRepository
	groupRepo *repository.GroupRepository
	scraper   *apple.AppleScraper
}

func NewAddAppleAppCommand(cfg *config.Config, adminRepo *repository.AdminRepository, groupRepo *repository.GroupRepository, scraper *apple.AppleScraper) *AddAppleAppCommand {
	return &AddAppleAppCommand{cfg: cfg, adminRepo: adminRepo, groupRepo: groupRepo, scraper: scraper}
}

func (c *AddAppleAppCommand) Execute(msg *tgbotapi.Message, sender Sender) {
	if !authorizeGroup(msg.Chat.ID, c.adminRepo, sender) {
		return
	}
	args := splitArgs(msg.CommandArguments())
	if len(args) < 1 || len(args) > 2 {
		_ = sender.SendMessage(msg.Chat.ID, "Invalid arguments")
		return
	}
	country := "vn"
	if len(args) == 2 {
		country = args[1]
	}

	// Java: try parsing arg[0] as Long (trackId); else treat as bundleId.
	var req request.AppleAppRequest
	if trackID, err := strconv.ParseInt(args[0], 10, 64); err == nil {
		req = request.ByTrackID(trackID, country)
	} else {
		req = request.ByBundleID(args[0], country)
	}

	resp, err := c.scraper.FetchAndCache(req)
	if err != nil || resp == nil || resp.AppID == "" {
		_ = sender.SendMessage(msg.Chat.ID, "Error when request app info")
		return
	}

	added, err := c.groupRepo.AddAppleApp(msg.Chat.ID, resp.AppID, country)
	if err != nil {
		_ = sender.SendMessage(msg.Chat.ID, "Internal server error")
		return
	}
	if !added {
		_ = sender.SendMessage(msg.Chat.ID, fmt.Sprintf("Apple app <code>%s</code> is already added", resp.AppID))
		return
	}
	_ = sender.SendMessage(msg.Chat.ID, fmt.Sprintf("Apple app <code>%s</code>, country <b>%s</b> added successfully", resp.AppID, country))
}

File: src/bot/commands/add-apple-app.js
import { buildAppleRequestByBundleId, buildAppleRequestByTrackId } from '../../api/apple-scraper.js';
import * as groupRepo from '../../repository/group-repository.js';
import { authorizeGroup, getCommandArguments, splitArgs } from './command-utils.js';

// /addapple <id|appId> [country=vn] — Java AddAppleAppCommand.
export function createAddAppleAppCommand(appleScraper) {
  return async (msg, sender) => {
    if (!(await authorizeGroup(msg.chat.id, sender))) return;
    const args = splitArgs(getCommandArguments(msg.text));
    if (args.length < 1 || args.length > 2) {
      await sender.sendMessage(msg.chat.id, 'Invalid arguments');
      return;
    }
    const country = args.length === 2 ? args[1] : 'vn';

    // Java: try parsing arg[0] as Long (trackId); else treat as bundleId.
    const trackId = Number.parseInt(args[0], 10);
    const req =
      Number.isFinite(trackId) && String(trackId) === args[0]
        ? buildAppleRequestByTrackId(trackId, country)
        : buildAppleRequestByBundleId(args[0], country);

    let resp;
    try {
      resp = await appleScraper.fetchAndCache(req);
    } catch {
      resp = null;
    }
    if (!resp || !resp.appId) {
      await sender.sendMessage(msg.chat.id, 'Error when request app info');
      return;
    }

    try {
      const added = await groupRepo.addAppleApp(msg.chat.id, resp.appId, country);
      if (!added) {
        await sender.sendMessage(msg.chat.id, `Apple app <code>${resp.appId}</code> is already added`);
        return;
      }
      await sender.sendMessage(
        msg.chat.id,
        `Apple app <code>${resp.appId}</code>, country <b>${country}</b> added successfully`,
      );
    } catch {
      await sender.sendMessage(msg.chat.id, 'Internal server error');
    }
  };
}
